### DISCUSSION TITLE
Proper way to use the hibernate api inside keycloak event listener

### DISCUSSION THREAD

**Discussion by shayanarm** [keycloak/keycloak#13800] [Q&A] [answered]
I have a running Keycloak instance backed by a Postgres database. My intention is to use exposed resource management endpoints from the available ```KeycloakSession``` object inside my event listener provider to query and manipulate some table records after certain events.

    import org.keycloak.models.KeycloakSession
    import org.keycloak.events.EventListenerProvider
    import org.keycloak.events.Event
    import org.keycloak.events.admin.AdminEvent
    import org.keycloak.models.AbstractKeycloakTransaction
    import org.keycloak.events.admin.OperationType
    import scala.jdk.CollectionConverters.*

    class MyEventListenerProvider(session: KeycloakSession)
        extends EventListenerProvider {
    
      object paths:
        val users = "users/(.+)".r // simplified for brevity
        val groups = "groups/(.+)".r // simplified for brevity
    
      override def onEvent(event: Event) = ()
    
      override def onEvent(
          adminEvent: AdminEvent,
          includeRepresentation: Boolean
      ) =
        session
          .getTransactionManager()
          .enlist(new AbstractKeycloakTransaction {
            protected def commitImpl(): Unit = {
              (adminEvent.getOperationType(), adminEvent.getResourcePath()) match
                case (OperationType.CREATE, paths.groups(id)) =>
                  val realm = session.realms().getRealm(adminEvent.getRealmId())
                  val group = session.groups().getGroupById(id, realm)
                  // The line below fails WITH a wrapping transaction consistently
                  // But works when it's not wrapped in a transaction
                  // Error message: no transaction is in progress!
                  List("provider", "admin").foreach { subgroup =>
                    session.groups().createGroup(realm, subgroup, group)
                  }
                case (OperationType.UPDATE, paths.users(id)) =>
                  val realm = session.realms().getRealm(adminEvent.getRealmId())
                  val user = session.users().getUserById(realm, id)
                  // The line below fails WITH a wrapping transaction for a
                  // couple of times before it succeeds after keycloak startup.
                  // It also fails consistently WITHOUT a wrapping transaction
                  // Error message: ResultSet is closed!
                  user.getGroups().asScala.toSeq.map(_.getName())
                  // Rest of the code does not matter; it fails at the line above
                case _ => ()
            }
    
            protected def rollbackImpl(): Unit = ()
          })
    
      override def close() = ()
    }

The errors I receive vary depending on which variation of the code above I run as explained by the inline code comments.

My running Keycloak is version ```18.0.0``` and here is the list of my dependencies:

    val KeycloakVersion = "18.0.0"

    libraryDependencies ++= Seq(
      "org.keycloak" % "keycloak-core" % KeycloakVersion % Provided,
      "org.keycloak" % "keycloak-server-spi" % KeycloakVersion % Provided,
      "org.keycloak" % "keycloak-services" % KeycloakVersion % Provided,
      "org.keycloak" % "keycloak-saml-core-public" % KeycloakVersion % Provided,
      "org.keycloak" % "keycloak-server-spi-private" % KeycloakVersion % Provided
    )

**Comment by lexcao** [comment, marked as answer]
I think you may not need to add transactions manually.
Could you please try to remove this code?
```
session
      .getTransactionManager()
      .enlist(new AbstractKeycloakTransaction {
```

Just to call
```
session.groups().createGroup(realm, subgroup, group)
```

And it would wrap to the current transaction automatically.

If you want more details, please refer to this
https://github.com/keycloak/keycloak/blob/ec4b1c01584b3b0065306c78afb8d77f797baa5b/model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java#L609
